File: src/com/systemvi/examples/minecraft/world/World.java
package com.systemvi.examples.minecraft.world;

import com.systemvi.engine.camera.CameraController2;
import com.systemvi.engine.noise.Perlin2d;
import com.systemvi.examples.minecraft.renderer.BlockFaceRenderer;
import com.systemvi.engine.camera.CameraController;
import org.joml.Vector3i;

public class World {
    private Chunk[][][] chunks;

    private Perlin2d noise;

    public World(){
        noise=new Perlin2d((int)System.currentTimeMillis(),100,100);
        chunks=new Chunk[15][2][15];
        for(int i=0;i<chunks.length;i++){
            for(int j=0;j<chunks[0].length;j++){
                for(int k=0;k<chunks[0][0].length;k++){
                    chunks[i][j][k]=new Chunk(new Vector3i(i,j,k),noise);
                }
            }
        }
        for(int i=0;i<chunks.length;i++){
            for(int j=0;j<chunks[0].length;j++){
                for(int k=0;k<chunks[0][0].length;k++){
                    chunks[i][j][k].generateCache(new Vector3i(i,j,k),this);
                }
            }
        }
    }

    public Chunk[][][] getChunks() {
        return chunks;
    }

    public BlockState getBlockState(int x, int y, int z){
        int chunkX=x/Chunk.SIZE_X;
        int chunkY=y/Chunk.SIZE_Y;
        int chunkZ=z/Chunk.SIZE_Z;
        x=x%Chunk.SIZE_X;
        y=y%Chunk.SIZE_Y;
        z=z%Chunk.SIZE_Z;
        if(chunkX<0||chunkX>=chunks.length||x<0)return new BlockState(Block.AIR);
        if(chunkY<0||chunkY>=chunks[0].length||y<0)return new BlockState(Block.AIR);
        if(chunkZ<0||chunkZ>=chunks[0][0].length||z<0)return new BlockState(Block.AIR);

        return chunks[chunkX][chunkY][chunkZ].blockStates[x][y][z];
    }

    public BlockState getBlockState(Vector3i pos){
        return getBlockState(pos.x, pos.y, pos.z);
    }

    public void setBlockState(BlockState state,int chunkX,int chunkY, int chunkZ,int x,int y,int z){
        chunks[chunkX][chunkY][chunkZ].blockStates[x][y][z]=state;
    }
}

File: voxel-world/src/scala/com/systemvi/voxel/world/world2/World.scala
package com.systemvi.voxel.world.world2

import com.systemvi.engine.texture.{Texture, TextureRegion}
import com.systemvi.voxel.world.generators.WorldGenerator
import com.systemvi.voxel.world.world.{Block, BlockState}
import org.joml.Vector3i


class World(val numberOfChunks: Vector3i) {

  val texture = Texture("assets/examples/minecraft/textures/diffuse.png")
  val regions: Array[Array[TextureRegion]] = TextureRegion.split(texture, 16, 16)
  Block.AIR = new Block(null, null, null, false)
  Block.STONE = new Block(regions(7)(0), regions(7)(0), regions(7)(0), true)
  Block.DIRT = new Block(regions(2)(3), regions(1)(2), regions(1)(3), true)

  val chunks: Array[Array[Array[Chunk]]] = Array.ofDim(numberOfChunks.x, numberOfChunks.y, numberOfChunks.z)

  def getBlockState(worldPosition: Vector3i): BlockState = getBlockState(
    worldPosition.x / Chunk.size.x,
    worldPosition.y / Chunk.size.y,
    worldPosition.z / Chunk.size.z,
    worldPosition.x % Chunk.size.x,
    worldPosition.y % Chunk.size.y,
    worldPosition.z % Chunk.size.z,
  )

  def getBlockState(chunkPosition: Vector3i, blockPosition: Vector3i): BlockState = {
    if (
      chunkPosition.x < 0 || chunkPosition.x >= numberOfChunks.x ||
        chunkPosition.y < 0 || chunkPosition.y >= numberOfChunks.y ||
        chunkPosition.z < 0 || chunkPosition.z >= numberOfChunks.z ||
        blockPosition.x < 0 || blockPosition.x >= Chunk.size.x ||
        blockPosition.y < 0 || blockPosition.y >= Chunk.size.y ||
        blockPosition.z < 0 || blockPosition.z >= Chunk.size.z
    )
      BlockState(Block.AIR)
    else
      chunks(chunkPosition.x)(chunkPosition.y)(chunkPosition.z)
        .blockStates(blockPosition.x)(blockPosition.y)(blockPosition.z)
  }

  def getBlockState(x: Int, y: Int, z: Int): BlockState = getBlockState(
    x / Chunk.size.x,
    y / Chunk.size.y,
    z / Chunk.size.z,
    x % Chunk.size.x,
    y % Chunk.size.y,
    z % Chunk.size.z,
  )

  def getBlockState(chunkX: Int, chunkY: Int, chunkZ: Int, blockX: Int, blockY: Int, blockZ: Int): BlockState = {
    if (
      chunkX < 0 || chunkY < 0 || chunkZ < 0 ||
        chunkX >= numberOfChunks.x || chunkY >= numberOfChunks.y || chunkZ >= numberOfChunks.z ||
        blockX < 0 || blockY < 0 || blockZ < 0 ||
        blockX >= Chunk.size.x || blockY >= Chunk.size.y || blockZ >= Chunk.size.z
    )
      BlockState(Block.AIR)
    else
      chunks(chunkX)(chunkY)(chunkZ).blockStates(blockX)(blockY)(blockZ)
  }


  def generate(generator: WorldGenerator): Unit = {
    for (i <- chunks.indices; j <- chunks(0).indices; k <- chunks(0)(0).indices) {
      chunks(i)(j)(k) = Chunk()
      val chunk = chunks(i)(j)(k)
      chunk.generate(generator = generator, chunkPosition = Vector3i(i, j, k))
    }
  }
}
